[Merged by Bors] - feat: `List.destutter` on cotransitive relations.

Timeroot opened the pull request:
`List.destutter` behaves nicely when the condition is a _cotransitive_ relation, and even better when it's a _coequivalence_ relation. This proves that behavior.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Part of #9052.

ericrbg commented on Mathlib/Data/List/Destutter.lean:
```diff
@@ -61,6 +63,15 @@ theorem destutter'_singleton : [b].destutter' R a = if R a b then [a, b] else [a
   split_ifs with h <;> simp! [h]
 #align list.destutter'_singleton List.destutter'_singleton
 
+/-- Destutter' always has length at least 1. -/
```
```suggestion
/-- `destutter'` always has length at least 1. -/
```
and similar through the file

ericrbg commented on Mathlib/Data/List/Destutter.lean:
```diff
@@ -61,6 +63,15 @@ theorem destutter'_singleton : [b].destutter' R a = if R a b then [a, b] else [a
 #align list.destutter'_singleton List.destutter'_singleton
 
+/-- Destutter' always has length at least 1. -/
+theorem destutter'_length_pos : 0 < (destutter' R a l).length := by
```
```suggestion
theorem length_destutter'_pos : 0 < (destutter' R a l).length := by
```

ericrbg commented on Mathlib/Data/List/Destutter.lean:
```diff
@@ -173,4 +184,185 @@ theorem destutter_eq_nil : ∀ {l : List α}, destutter R l = [] ↔ l = []
   | _ :: l => ⟨fun h => absurd h <| l.destutter'_ne_nil R, fun h => nomatch h⟩
 #align list.destutter_eq_nil List.destutter_eq_nil
 
+/--Destutter of a map is the same as the map of the destutter, as long as the map preserves the
+   relation. -/
```
```suggestion
/-- For a relation-preserving map, `destutter'` commutes with `map`. -/
```

@YaelDillies, should there be a version for `RelHom`?

YaelDillies replied:
Probably not. We should just state this for injective maps because I doubt we care about any other relation than `≠`.

ericrbg commented on Mathlib/Data/List/Destutter.lean:
```diff
@@ -173,4 +184,185 @@ theorem destutter_eq_nil : ∀ {l : List α}, destutter R l = [] ↔ l = []
+    In other words: (l.destutter R) is an R-chain sublist of l;
+    it is at least as long as any other R-chain sublist.
+
+Proof sketch:
```
This is great! I think the docstring is the wrong place for it - I'd sprinkle it in through the proof.

ericrbg reviewed (commented):
I've added the "proper" depending relation - the bot should now do a little bit of work for you!

A few comments, I think the rest of the docstrings should be edited in similar ways too.

Thanks for all this! Contributing to Mathlib for the first time is hard and long sometimes, but we really appreciate it :)

cc @YaelDillies so you can see where `IsCoequiv` is used

Timeroot commented:
Alright, I dropped the "Coequivalence" and "Cotransitive" stuff; now everything uses `compl`. I added an instance for `NEq` to get the right stuff going. Also adding the `destutter_map_inj` for injective maps.

Timeroot commented:
I removed the dependence on #9081 since it's not using any of that code now.

YaelDillies commented on Mathlib/Data/List/Destutter.lean:
```diff
@@ -200,43 +199,49 @@ theorem destutter_map_iff (f : α → β) (h : ∀ a b, R a b ↔ R₂ (f a) (f
 
-/-- Destutter' on a relation like ≠, whose negation is an equivalence, gives the same length if
+/-- For a injective function `f`, `destutter' (·≠·)` commutes with `map f`. -/
+theorem destutter_map_inj {f : α → β} (h : Function.Injective f) [DecidableEq α] [DecidableEq β] :
```
```suggestion
theorem map_destutter_ne {f : α → β} (h : Function.Injective f) [DecidableEq α] [DecidableEq β] :
```
Also feel free to `open Function` at the top to be able to write
```suggestion
theorem map_destutter_ne {f : α → β} (h : Injective f) [DecidableEq α] [DecidableEq β] :
```

YaelDillies commented on Mathlib/Data/List/Destutter.lean:
```diff
@@ -200,43 +199,49 @@ theorem destutter_map_iff (f : α → β) (h : ∀ a b, R a b ↔ R₂ (f a) (f
+/-- `destutter'` on a relation like ≠, whose negation is an equivalence, gives the same length if
     the first elements are not related (¬Rα a b). --/
-theorem length_destutter'_eq [IsCoequiv α R] (hab : ¬R a b) :
+theorem length_destutter'_eq [IsEquiv α Rᶜ] (hab : ¬R a b) :
```
```suggestion
theorem length_destutter' [IsEquiv α Rᶜ] (hab : ¬R a b) :
```

YaelDillies commented on Mathlib/Data/List/Destutter.lean:
```diff
@@ -200,43 +199,49 @@ theorem destutter_map_iff (f : α → β) (h : ∀ a b, R a b ↔ R₂ (f a) (f
     monotonic under List.cons --/
-theorem length_destutter'_ge_length_destutter'_cons [IsCoequiv α R] :
+theorem length_destutter'_ge_length_destutter'_cons [IsEquiv α Rᶜ] :
     (List.destutter' R a (b::l)).length ≥ (List.destutter' R b l).length := by
```
```suggestion
theorem le_length_destutter' [IsEquiv α Rᶜ] :
    (List.destutter' R b l).length ≤ (List.destutter' R a (b :: l)).length := by
```

Timeroot replied:
Swapped this and similar inequalities

YaelDillies commented on Mathlib/Data/List/Destutter.lean:
```diff
@@ -200,43 +199,49 @@ theorem destutter_map_iff (f : α → β) (h : ∀ a b, R a b ↔ R₂ (f a) (f
     by_cases hab : (R a b)
     case pos => simp [destutter', if_pos hab, Nat.le_succ]
-    by_cases hac : (R a c)
+    by_cases hac : (Rᶜ a c)
```
```suggestion
    by_cases hac : R a c
```
and swap the two branches

Timeroot replied:
This actually doesn't easily work. Using the `Rᶜ a c` is a deliberate choice because it helps the typeclass inference realize that we're talking about `Rᶜ` instead of just `R`, and lets it find the relevant `trans Rᶜ`. With `by_cases hac : R a c`, the `neg` branch gets `¬R a c` which doesn't work with the typeclass inference like `Rᶜ a c` does. Of course you can get around this but it adds several lines of noise. (Maybe there's a super-easy transparent way that I'm missing.)

YaelDillies replied:
You can always do
```
obtain hac | hac : R a c ∨ Rᶜ a c := em _
```

Timeroot replied:
Alrighty, changed this

YaelDillies commented on Mathlib/Data/List/Destutter.lean:
```diff
@@ -212,13 +212,13 @@ theorem length_destutter' [IsEquiv α Rᶜ] (hab : ¬R a b) :
+      have hbc : ¬Rᶜ b c := mt (_root_.trans hab) (Classical.not_not.2 hac)
+      simp [destutter', if_pos hac, if_pos (Classical.not_not.1 hbc)]
+    next =>
       have hbc : ¬R b c := trans (symm hab) hac
```
Why not just
```suggestion
    · have hbc : ¬Rᶜ b c := mt (_root_.trans hab) (Classical.not_not.2 hac)
      simp [destutter', if_pos hac, if_pos (Classical.not_not.1 hbc)]
    · have hbc : ¬R b c := trans (symm hab) hac
```

Timeroot replied:
Done, also dropped an extra couple `Classical.`s

YaelDillies commented on Mathlib/Data/List/Basic.lean:
```diff
@@ -4382,6 +4416,13 @@ theorem getD_replicate_default_eq (r n : ℕ) : (replicate r d).getD n d = d :=
 #align list.nthd_replicate_default_eq List.getD_replicate_default_eqₓ -- argument order
 
+@[simp]
+theorem getD_replicate_elem_eq {y} (i n) (h : i < n) :
```
```suggestion
theorem getD_replicate {y i n} (h : i < n) :
```

YaelDillies commented on Mathlib/Data/List/Destutter.lean:
```diff
@@ -61,6 +64,15 @@ theorem destutter'_singleton : [b].destutter' R a = if R a b then [a, b] else [a
 #align list.destutter'_singleton List.destutter'_singleton
 
+/-- `destutter'` always has length at least 1. -/
+theorem length_destutter'_pos : 0 < (destutter' R a l).length := by
```
```suggestion
theorem destutter'_ne_nil : destutter' R a l ≠ [] = by
```
is the more canonical spelling

Timeroot replied:
Alright! So it turns out `destutter'_ne_nil` already exists so I just changed my one usage,
`exact length_destutter'_pos _ _`
into
`List.length_pos_of_ne_nil (destutter'_ne_nil _ _)`

YaelDillies commented on Mathlib/Data/List/Destutter.lean:
```diff
@@ -173,4 +185,187 @@ theorem destutter_eq_nil : ∀ {l : List α}, destutter R l = [] ↔ l = []
 #align list.destutter_eq_nil List.destutter_eq_nil
 
+/-- For a relation-preserving map, `destutter'` commutes with `map`. -/
+theorem destutter_map_iff (f : α → β) (h : ∀ a b, R a b ↔ R₂ (f a) (f b)) :
```
```suggestion
theorem map_destutter (f : α → β) (h : ∀ a b, R a b ↔ R₂ (f a) (f b)) :
```

YaelDillies commented on Mathlib/Data/List/Destutter.lean:
```diff
@@ -173,4 +185,187 @@ theorem destutter_eq_nil : ∀ {l : List α}, destutter R l = [] ↔ l = []
+/-- For a injective function `f`, `destutter' (·≠·)` commutes with `map f`. -/
+theorem map_destutter_ne {f : α → β} (h : Injective f) [DecidableEq α] [DecidableEq β] :
+    (l.destutter (·≠·)).map f = (l.map f).destutter (·≠·) :=
+  destutter_map_iff l _ f (fun _ _ ↦ (Injective.ne_iff h).symm)
```
```suggestion
  map_destutter l _ f fun _ _ ↦ h.ne_iff.symm
```

YaelDillies commented on Mathlib/Data/List/Destutter.lean:
```diff
@@ -173,4 +185,187 @@ theorem destutter_eq_nil : ∀ {l : List α}, destutter R l = [] ↔ l = []
+    (l.destutter (·≠·)).map f = (l.map f).destutter (·≠·) :=
+  destutter_map_iff l _ f (fun _ _ ↦ (Injective.ne_iff h).symm)
+
+/-- `destutter'` on a relation like ≠, whose negation is an equivalence, gives the same length if
+    the first elements are not related (¬Rα a b). --/
```
Just FYI your `destutter'` won't turn into a link in the docs unless you "fully qualify" the name:
```suggestion
/-- `List.destutter'` on a relation like `≠`, whose negation is an equivalence, gives the same
length if the first elements are not related. -/
```
Also, docstrings end with `-/`, not `--/`. Be careful! Else you'll end up with docstrings with a trailing `-`.

YaelDillies commented on Mathlib/Data/List/Destutter.lean:
```diff
@@ -173,4 +185,187 @@ theorem destutter_eq_nil : ∀ {l : List α}, destutter R l = [] ↔ l = []
+/-- `destutter'` on a relation like ≠, whose negation is an equivalence, gives the same length if
+    the first elements are not related (¬Rα a b). --/
+theorem length_destutter' [IsEquiv α Rᶜ] (hab : ¬R a b) :
+    (List.destutter' R a l).length = (List.destutter' R b l).length := by
```
```suggestion
    (destutter' R a l).length = (destutter' R b l).length := by
```
or
```suggestion
    (l.destutter' R a).length = (l.destutter' R b).length := by
```

YaelDillies commented on Mathlib/Data/List/Destutter.lean:
```diff
@@ -173,4 +185,187 @@ theorem destutter_eq_nil : ∀ {l : List α}, destutter R l = [] ↔ l = []
+
+/-- `destutter'` on a relation like ≠, whose negation is an equivalence, gives the same length if
+    the first elements are not related (¬Rα a b). --/
+theorem length_destutter' [IsEquiv α Rᶜ] (hab : ¬R a b) :
```
```suggestion
theorem length_destutter'_congr [IsEquiv α Rᶜ] (hab : ¬R a b) :
```
sounds more accurate

YaelDillies commented on Mathlib/Data/List/Destutter.lean:
```diff
@@ -173,4 +185,187 @@ theorem destutter_eq_nil : ∀ {l : List α}, destutter R l = [] ↔ l = []
+
+/-- `destutter'` on a relation like ≠ or <, whose negation is transitive, has length monotonic
+    under a ¬R changing of the first element. -/
+theorem length_destutter'_cotrans_ge [i : IsTrans α Rᶜ] (hba : ¬R b a) :
+    (List.destutter' R b l).length ≤ (List.destutter' R a l).length := by
```
Does this used twice not imply ` length_destutter'` straightforwardly?

Timeroot replied:
It is, good point. Changed

YaelDillies commented on Mathlib/Data/List/Destutter.lean:
```diff
@@ -173,4 +185,187 @@ theorem destutter_eq_nil : ∀ {l : List α}, destutter R l = [] ↔ l = []
+  · exact le_length_destutter' _ R
+
+/-- `destutter ≠` has length monotonic under List.cons --/
+theorem length_destutter_ne_cons_ge_length_destutter [DecidableEq α]:
```
```suggestion
theorem length_destutter_ne_cons_ge_length_destutter [DecidableEq α] :
```

YaelDillies commented on Mathlib/Data/List/Destutter.lean:
```diff
@@ -173,4 +185,187 @@ theorem destutter_eq_nil : ∀ {l : List α}, destutter R l = [] ↔ l = []
+
+/-- `destutter'` on a relation like ≠, whose negation is an equivalence, has length
+    monotonic under List.cons --/
+theorem le_length_destutter' [IsEquiv α Rᶜ] :
+    (List.destutter' R b l).length ≤ (List.destutter' R a (b :: l)).length := by
```
Can you generalise that to
```suggestion
theorem Suffix.length_destutter'_mono [IsEquiv α Rᶜ] (h : a :: l₁ <:+ b :: l₂) :
    (List.destutter' R a l₁).length ≤ (List.destutter' R b l₂).length := by
```

Timeroot replied:
I think that's a correct statement as written, but it would probably just be an induction with `le_length_destutter'` as the inductive step. I don't really know anything about `Suffix` and I'm not super motivated to dive into that atm, so I hope it's alright I skip this one.

YaelDillies replied:
Mind leaving 
```suggestion
/-
TODO: Replace this lemma by the more general version:
theorem Suffix.length_destutter'_mono [IsEquiv α Rᶜ] (h : a :: l₁ <:+ b :: l₂) :
    (List.destutter' R a l₁).length ≤ (List.destutter' R b l₂).length
-/
theorem le_length_destutter' [IsEquiv α Rᶜ] :
    (List.destutter' R b l).length ≤ (List.destutter' R a (b :: l)).length := by
```

YaelDillies commented on Mathlib/Data/List/Destutter.lean:
```diff
@@ -173,4 +185,187 @@ theorem destutter_eq_nil : ∀ {l : List α}, destutter R l = [] ↔ l = []
+    gives a list of maximal length over any chain. In other words: `l.destutter R` is an
+    R-chain sublist of l, and is at least as long as any other R-chain sublist.
+-/
+theorem length_destutter_coequiv_maximal_chain_neg_trans [IsEquiv α Rᶜ]
+    (h₁ : l₂ <+ l) (h₂ : l₂.Chain' R) :
+    l₂.length ≤ (l.destutter R).length := by
```
This is a completely unguessable name. Maybe
```suggestion
theorem Chain.length_le_length_destutter [IsEquiv α Rᶜ] (h₁ : l₂ <+ l) (h₂ : l₂.Chain' R) :
    l₂.length ≤ (l.destutter R).length := by
```

Timeroot replied:
Thanks. I have just barely enough intuition to guess names sometimes myself. I recognized this name was unguessable, but had no idea how to name it myself. :)

YaelDillies commented on Mathlib/Init/Algebra/Classes.lean:
```diff
@@ -1,7 +1,7 @@
 Copyright (c) 2017 Microsoft Corporation. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
-Authors: Leonardo de Moura
+Authors: Leonardo de Moura, Alex Meiburg
```
```suggestion
Authors: Leonardo de Moura
```

YaelDillies commented on Mathlib/Order/Basic.lean:
```diff
@@ -782,6 +782,17 @@ instance IsRefl.compl (r) [IsRefl α r] : IsIrrefl α rᶜ :=
   ⟨fun a ↦ not_not_intro (refl a)⟩
 #align is_refl.compl IsRefl.compl
 
+instance DecidableRel.compl (r : α → α → Prop) [i : DecidableRel r] : DecidableRel rᶜ :=
```
Is this not already inferred?

Timeroot replied:
Oh. I guess it is. I was struggling to get things with complements to work, and it seemed like this was necessary ... but I guess it isn't. Will cut it out.

YaelDillies commented on Mathlib/Order/Basic.lean:
```diff
@@ -782,6 +782,17 @@ instance IsRefl.compl (r) [IsRefl α r] : IsIrrefl α rᶜ :=
+      (fun h _ ↦ isTrue h)
+      (fun h _ ↦ isFalse (not_not_intro h)) rfl
+
+instance NEq.compl : IsEquiv α (·≠·)ᶜ := by
```
```suggestion
instance Ne.instIsEquiv_compl : IsEquiv α (· ≠ ·)ᶜ := by
```

YaelDillies commented on Mathlib/Order/Basic.lean:
```diff
@@ -782,6 +782,17 @@ instance IsRefl.compl (r) [IsRefl α r] : IsIrrefl α rᶜ :=
+      (fun h _ ↦ isFalse (not_not_intro h)) rfl
+
+instance NEq.compl : IsEquiv α (·≠·)ᶜ := by
+  have cnot : Notᶜ = id := funext fun _ ↦ propext not_not
+  simp_rw [Pi.compl_def, ne_eq, ← Pi.compl_apply, cnot]
+  exact eq_isEquiv α
```
Something like
```suggestion
  convert eq_isEquiv α; simp
```

Timeroot replied:
Turns out `simp [compl]` works.

YaelDillies commented:
I don't have time to review this any further but I was globally happy with it during my last review. So whoever takes over reviewing should only look for local issues.

urkud commented on Mathlib/Data/List/Basic.lean:
```diff
@@ -4382,6 +4415,13 @@ theorem getD_replicate_default_eq (r n : ℕ) : (replicate r d).getD n d = d :=
+@[simp]
+theorem getD_replicate {y i n} (h : i < n) :
+    getD (replicate n x) i y = x := by
+  rw [getD, get?_eq_get, get_replicate]
```
We have `getD_eq_get`

Timeroot replied:
done

urkud commented on Mathlib/Data/List/Destutter.lean:
```diff
@@ -173,4 +176,187 @@ theorem destutter_eq_nil : ∀ {l : List α}, destutter R l = [] ↔ l = []
+theorem Sublist.length_destutter'_mono [IsEquiv α Rᶜ] (h : a :: l₁ <+ b :: l₂) :
+    (List.destutter' R a l₁).length ≤ (List.destutter' R b l₂).length
+-/
+theorem le_length_destutter' [IsEquiv α Rᶜ] :
```
```suggestion
theorem le_length_destutter'_cons [IsEquiv α Rᶜ] :
```

Timeroot commented:
@urkud other stuff this needs?

Timeroot commented:
@YaelDillies do you know who else could review this? Other PRs are still blocked by it and it's been several months.

urkud commented:
I'm sorry for the delay. I'm going to merge `master` and fix compile now, then review it tonight or tomorrow.

YaelDillies commented:
Wow sorry, this completely fell off my radar. I guess I will wait for Yury to merge master, but yes we should get that in!

Timeroot commented:
Lightly bumping this @urkud :) If you don't have time, I can try to find time to update to master, but it won't be for a bit

urkud commented:
I'm sorry for the delay. I failed to finish it before traveling, then forgot to come back. I'm going to merge my local changes with @YaelDillies 's fixes today. I cherry-picked a couple of lemmas that are ready to get merged to a new PR.

urkud commented:
Asked a question about `List.destutter'` on Zulip: https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/List.2Edestutter'

Timeroot commented:
Is this good to merge now?

Timeroot commented:
Oh, I see the Zulip thread left this unresolved. I guess I'll wait for that then

YaelDillies commented:
I think you've waited long enough, so I pushed a little golf and LGTM

maintainer merge

kim-em commented on Mathlib/Data/List/Destutter.lean:
```diff
@@ -148,4 +151,116 @@ theorem destutter_eq_nil : ∀ {l : List α}, destutter R l = [] ↔ l = []
 
+variable {R}
+
+/-- For a relation-preserving map, `destutter'` commutes with `map`. -/
```
```suggestion
/-- For a relation-preserving map, `destutter'` commutes with `map`. -/
```
Doc-strings need updating to remove the prime.

YaelDillies replied:
I assume you mean the prime in `destutter'`? Your suggestion above doesn't seem to contain any change 😄

kim-em commented:
bors d+

Timeroot commented:
bors r+

kim-em commented:
I'm not sure what happened with bors here. Let's try again.

bors merge
